### PR TITLE
Committing changes for issue 657

### PR DESCRIPTION
**Pull request by Maheshwaran17**
Hi @christianp , I have removed the auto-close of the URL and made the remaining changes. Submitting this pull request for the [issue#657](https://github.com/numbas/editor/issues/657). Could you please approve this. Thanks in advance!

### CODE REVIEW

File: editor/static/css/preview.css [hunk 1]
```diff
@@ -12,18 +12,59 @@ header {
...
+
+.no-underline-link {
+    text-decoration:none;
+}
+
+.underline-text{
+    text-decoration: underline;
+}
+
+.pointer-effect{
```
Comment (christianp) [inline editor/static/css/preview.css:36]: This is a hack around the fact that you want tags with this class to work like buttons. The correct thing to do is to make them buttons, and style them to look like text.

File: editor/static/css/preview.css [hunk 2]
```diff
@@ -12,18 +12,59 @@ header {
...
+    cursor: pointer;
+}
+
+.light-maroon {
+    color: hsl(0, 31%, 79%);
```
Comment (christianp) [inline editor/static/css/preview.css:41]: I'm not sure where this colour came from, but I'd prefer to name the class after its meaning, rather than the colour it'll give.

File: editor/static/css/preview.css [hunk 3]
```diff
@@ -12,18 +12,59 @@ header {
...
+}
+
+.color-white {
+    color: hsl(0, 0%, 100%);
+}
+
+.copied-text-style{
```
Comment (christianp) [inline editor/static/css/preview.css:48]: I don't think `-style` is a necessary suffix for a class name!

File: editor/templates/editoritem/preview.html [hunk 4]
```diff
@@ -20,31 +20,35 @@
...
+                <span class="light-maroon">in</span>
+                <a href="{{item.project.get_absolute_url}}">{{item.project.name}}</a>
             </h1>
 
             <span class="licence">
             {% if item.licence %}
-                &copy;
-                <abbr title="{{item.licence.name}}">
+            <span class="light-maroon no-underline-link"> &copy; </span>
+              <!--  <abbr title="{{item.licence.name}}"> -->
```
Comment (christianp) [inline editor/templates/editoritem/preview.html:31]: why did you remove the `<abbr>` tag?

File: editor/templates/editoritem/preview.html [hunk 5]
```diff
@@ -20,31 +20,35 @@
...
-            <button class="btn" id="share" type="button">Share</button>
+            <span class="pointer-effect underline-text" id="share"> Share </span>
 
-            <span id="embed-url">
-                <code tabindex="0">{{SITE_ROOT}}{{embed_url}}</code>
+            <span  id="embed-url">
+                <span class="copied-text-style" id="copied-text"></span>
+                <span class="light-maroon" id="link-label">Click <span class="pointer-effect underline-text color-white" id="copy-text">here</span> to copy this link - </span>
+                <code tabindex="0"><span class="color-white">{{SITE_ROOT}}{{embed_url}}</span></code>
+                <span class="close-button pointer-effect" id="close-button">&nbsp Close &nbsp</span>
```
Comment (christianp) [inline editor/templates/editoritem/preview.html:51]: This is a button, so it should be a `<button>` tag.

File: editor/templates/editoritem/preview.html [hunk 6]
```diff
@@ -20,31 +20,35 @@
...
             <a href="{% editoritem_url 'download' item.rel_obj %}?scorm">Download</a>
 
-            <button class="btn" id="share" type="button">Share</button>
+            <span class="pointer-effect underline-text" id="share"> Share </span>
 
-            <span id="embed-url">
-                <code tabindex="0">{{SITE_ROOT}}{{embed_url}}</code>
+            <span  id="embed-url">
+                <span class="copied-text-style" id="copied-text"></span>
+                <span class="light-maroon" id="link-label">Click <span class="pointer-effect underline-text color-white" id="copy-text">here</span> to copy this link - </span>
```
Comment (christianp) [inline editor/templates/editoritem/preview.html:49]: Text like "click here" isn't great for screen readers - you'd have to navigate out of the button to find out what it will do. So the clickable text should be "Copy this link"

File: editor/templates/editoritem/preview.html [hunk 7]
```diff
@@ -54,19 +58,48 @@ <h1 class="description">
...
+            close_button.addEventListener('click',function() {
+                embed_display.classList.toggle('shown'); 
+                embed_display.focus();
+            });
+
+            /* Function to copy the URL to the clipboard */
+            copy_text.addEventListener('click',function() {
+                                
+                /* To copy the URL to clipboard */
+                navigator.clipboard.writeText(`{{SITE_ROOT}}{{embed_url}}`);
```
Comment (christianp) [inline editor/templates/editoritem/preview.html:83]: I'm wary of interpolating into javascript - odd characters could break the whole script. We already have the URL in the `<code>` tag, so fetch it from there.

File: editor/templates/editoritem/preview.html [hunk 8]
```diff
@@ -54,19 +58,48 @@ <h1 class="description">
...
+
+            var share_button = document.getElementById('share');
+            var embed_display = document.getElementById('embed-url');
+            var copy_text = document.getElementById('copy-text');
+            var close_button = document.getElementById('close-button');
+
+            /* To toggle the embedded URL area */
+            share_button.addEventListener('click',function() {
+                embed_display.classList.toggle('shown'); 
+                embed_display.focus();
```
Comment (christianp) [inline editor/templates/editoritem/preview.html:70]: The `#embed-url` tag doesn't appear in the tab order, so it can't be focused. You wouldn't want to focus it anyway - it's not interactive.

File: editor/templates/editoritem/preview.html (hunk 7 above)
Comment (christianp) [reply editor/templates/editoritem/preview.html:83]: And `navigator.clipboard.writeText` is a fairly new feature, so you should check it exists before calling it.

File: editor/templates/editoritem/preview.html [hunk 9]
```diff
@@ -54,19 +58,48 @@ <h1 class="description">
...
+            });
+
+            /* Function to copy the URL to the clipboard */
+            copy_text.addEventListener('click',function() {
+                                
+                /* To copy the URL to clipboard */
+                navigator.clipboard.writeText(`{{SITE_ROOT}}{{embed_url}}`);
+                
+                /* Displays the message that URL is copied to clipboard */
+                document.getElementById("copied-text").innerHTML = '<span class="left-right-padding">&nbsp URL Copied &nbsp</span>';
```
Comment (christianp) [inline editor/templates/editoritem/preview.html:86]: Why write HTML in JS, when you could have it already in the page and toggle its visibility with CSS?

File: editor/static/css/preview.css [hunk 10]
```diff
@@ -12,18 +12,59 @@ header {
     /*! border-bottom: 0.2rem solid hsl(0, 70%, 10%); */
     display: grid;
     color: white;
-    box-shadow: 0 0px 1px 2px hsl(0, 70%, 10%);
     z-index: 100;
-    line-height: 1.5em;
+    line-height: 0.5em;
```
Comment (christianp) [inline editor/static/css/preview.css:16]: line-height shouldn't be less than 1em - it will lead to things overlapping when you don't expect them to. This happened when I made the screen very narrow. I think you really wanted to change the padding or margins around some elements.

**Review by christianp — COMMENTED** [review]
This looks good, thanks! 
I've made some comments on things that I'd do differently. I'm going to commit some changes that resolve these, but I'm leaving these comments for your benefit.
The most important things are that you shouldn't try to replicate button semantics - just use buttons - and to be aware of how screen readers interact with the page.

**Comment by Maheshwaran17** [comment]
@christianp Thank you very much for reviewing the code and also for your valuable comments which would help me in future :)  

Please let me know if i need to make any changes and resubmit the code from my side.
